cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -102,12 +102,12 @@ type upgrader struct {
 	initFn                    func() error
 	printIntroHelpTextFn      func() error
 	validateAndDeriveParamsFn func() error
-	prepareFn                 func() (*daisy.Workflow, error)
-	upgradeFn                 func() (*daisy.Workflow, error)
-	retryUpgradeFn            func() (*daisy.Workflow, error)
-	rebootFn                  func() (*daisy.Workflow, error)
-	cleanupFn                 func() (*daisy.Workflow, error)
-	rollbackFn                func() (*daisy.Workflow, error)
+	prepareFn                 func() (daisyutils.DaisyWorker, error)
+	upgradeFn                 func() (daisyutils.DaisyWorker, error)
+	retryUpgradeFn            func() (daisyutils.DaisyWorker, error)
+	rebootFn                  func() (daisyutils.DaisyWorker, error)
+	cleanupFn                 func() (daisyutils.DaisyWorker, error)
+	rollbackFn                func() (daisyutils.DaisyWorker, error)
```

Comment:
Good question. The interface is implemented with pointer receivers, so the object that's being passed around is a pointer to the struct that implements the interface:

https://github.com/GoogleCloudPlatform/compute-image-tools/blob/85f7e15dfd425f494df6c7353953d92cc925e6fb/cli_tools/common/utils/daisyutils/daisy_worker.go#L62

I think this is normal across Go's typical usage; eg in the standard library, the functions that accept implementations of Reader and Writer take the interfaces directly rather than as pointers (which would be a pointer to a pointer to a struct that implements that interface)

https://pkg.go.dev/io#Copy

I'm still kind of new to Go, so my understanding could be wrong here! But that was the intent.

Reply:
I missed the fact `DaisyWorker` is an interface. Nevermind.